new-tutorial/L1_large_node_classification.py

```diff
@@ -0,0 +1,326 @@
+# each node to aggregate messages.
+# 
+# The syntax of ``dgl.dataloading.NodeDataLoader`` is mostly similar to a
+# PyTorch ``DataLoader``, with the addition that it needs a graph to
+# generate computation dependency from, a set of node IDs to iterate on,
+# and the neighbor sampler you defined.
+# 
+# Let’s say that each node will gather message from 4 neighbors on each
+# layer. The code defining the data loader and neighbor sampler will look
+# like the following.
+# 
+
+sampler = dgl.dataloading.MultiLayerNeighborSampler([4, 4])
+train_dataloader = dgl.dataloading.NodeDataLoader(
+    graph, train_nids, sampler,
+    batch_size=1024,
+    shuffle=True,
+    drop_last=False,
+    num_workers=0,
+    device='cuda'
+)
+
+
+######################################################################
+# You can iterate over the data loader and see what it yields.
+# 
+
+input_nodes, output_nodes, bipartites = example_minibatch = next(iter(train_dataloader))
+print(example_minibatch)
+print("To compute {} nodes' outputs, we need {} nodes' input features".format(len(output_nodes), len(input_nodes))) 
+
+
+######################################################################
+# ``NodeDataLoader`` gives us three items per iteration.
+# 
+# -  The input node list for the nodes whose input features are needed to
+#    compute the outputs.
+# -  The output node list whose GNN representation are to be computed.
+# -  The list of computation dependency for each layer as a list of
+#    bipartite graphs.
+# 
+
+
+######################################################################
+# You can get the input and output node IDs of the first bipartite graph
+# and see that the first few input nodes are always the same as the output
+# nodes.
+# 
+
+bipartite_0_src = bipartites[0].srcdata[dgl.NID]
+bipartite_0_dst = bipartites[0].dstdata[dgl.NID]
+print(bipartite_0_src)
+print(bipartite_0_dst)
+print(torch.equal(bipartite_0_src[:bipartites[0].num_dst_nodes()], bipartite_0_dst))
+
+
+######################################################################
+# Defining Model
+# --------------
+# 
+# Let’s consider training a 2-layer GraphSAGE with neighbor sampling. The
+# model can be written as follows:
+# 
+
+import torch.nn as nn
+import torch.nn.functional as F
+from dgl.nn import SAGEConv
+
+class Model(nn.Module):
+    def __init__(self, in_feats, h_feats, num_classes):
+        super(Model, self).__init__()
+        self.conv1 = SAGEConv(in_feats, h_feats, aggregator_type='mean')
+        self.conv2 = SAGEConv(h_feats, num_classes, aggregator_type='mean')
+        self.h_feats = h_feats
+        
+    def forward(self, bipartites, x):
+        h_dst = x[:bipartites[0].num_dst_nodes()]  # <---
+        h = self.conv1(bipartites[0], (x, h_dst))  # <---
+        h = F.relu(h)
+        h_dst = h[:bipartites[1].num_dst_nodes()]  # <---
+        h = self.conv2(bipartites[1], (h, h_dst))  # <---
```

Comment:
What are those "# <---" for?

Reply:
As Tong said, they highlight the changes.  I added a comment.